mongoengine/document.py

```diff
@@ -1036,3 +1038,27 @@ def object(self):
             self._key_object = self._document.objects.with_id(self.key)
             return self._key_object
         return self._key_object
+
+
+class OrderedDocument(Document):
+    """A document that is almost same with Document except for returning
+    results in OrderedDict instead of dict.
+    """
+
+    # The __metaclass__ attribute is removed by 2to3 when running with Python3
+    # my_metaclass is defined so that metaclass can be queried in Python 2 & 3
+    my_metaclass = TopLevelDocumentMetaclass
+    __metaclass__ = TopLevelDocumentMetaclass
+
+    @classmethod
+    def _get_collection(cls):
+        collection = super(OrderedDocument, cls)._get_collection()
+
+        if IS_PYMONGO_3:
+            # returns collection object which is set OrderedDict class to be decoded from BSON document
+            from bson import CodecOptions
+            return collection.with_options(codec_options=CodecOptions(document_class=OrderedDict))
+        else:
+            # set attribute to specify the class to be decoeded
+            cls.decoded_class = OrderedDict
```

Comment:
This looks dangerous - you're setting a public property on the class within a specific condition of a particular method. Could it be set in an overridden `__init__` instead?